Implement version-sort for imports in `style_edition` 2024

ytmimi opened the pull request:
The `version-sort` algorithm is described in the [sorting section of the style guide] and the tracking issue for this in the `r-l/rust` repo is https://github.com/rust-lang/rust/issues/123800.

[sorting section of the style guide]: https://doc.rust-lang.org/nightly/style-guide/#sorting

This PR implements the algorithm and applies it to import sorting. One of the original PRs that implement this feature was #3764. In that PR the sort order for type aliases was also updated. I decided to keep this PR focused on import sorting, and I plan to open up follow up PRs to address any other sorting that rustfmt needs to apply to conform with the 2024 style_edition.

Given the nature of changing the sort order there were a lot of imports that needed to change in order to pass our self tests. Those changes have all been broken out into their own commit so you can review those changes in isolation from the implementation work. Overall, I've broken the PR into what I think are logical commits and I think it would be best to review the PR one commit at a time.

r? @calebcartwright

calebcartwright commented:
also cc @rust-lang/style for awareness as the diff of rustfmt's own imports should demonstrate the effects of the sorting changes

ytmimi commented on src/sort.rs:
```diff
@@ -0,0 +1,366 @@
+                (VersionChunk::Underscore, _) => return std::cmp::Ordering::Less,
+                (_, VersionChunk::Underscore) => return std::cmp::Ordering::Greater,
+                (VersionChunk::Str(ca), VersionChunk::Str(cb))
+                | (VersionChunk::Str(ca), VersionChunk::Number { source: cb, .. })
+                | (VersionChunk::Number { source: ca, .. }, VersionChunk::Str(cb)) => {
```
I think there's an opportunity to slightly optimize this. When comparing a `VersionChunk::Str` with a `VersionChunk::Number` I believe that the number will always get sorted first, but I implemented it this way because of this language in the style guide:

> To compare two chunks if both are not numeric, compare them by Unicode character lexicographically

calebcartwright commented on src/sort.rs:
```diff
@@ -0,0 +1,366 @@
+    fn parse_numeric_chunk(
+        &mut self,
+        mut chars: std::str::CharIndices<'a>,
+    ) -> Option<VersionChunk<'a>> {
```
No changes requested, but sharing an observation that we've got a lot of overlapping code between `parse_numeric_chunk` and `parse_str_chunk`, with the only difference (other than the final return value) being the couple checks in the while loop.

There's probably an opportunity to DRY that up in the future

ytmimi replied:
Thanks for taking a look. I'll address those merge conflicts and rebase the PR sometime later this week

calebcartwright commented on src/sort.rs:
```diff
@@ -0,0 +1,366 @@
+        input.sort_by(|a, b| version_sort(a, b));
+        assert_eq!(input, expected);
+
+        let mut input = vec![
+            "x87", "Z_YWX", "ZY_WX", "ZYW_X", "ZYWX", "ZYWX_", "v010", "v10", "w005s09t",
+            "w5s009t", "x64", "x86", "x86_32", "x86_64", "x86_128", "v000", "v00", "v0", "v0s",
+            "v00t", "v0u", "v001", "v01", "v1", "v009", "v09", "v9", "_ZYWX", "u_zzz", "u8", "u16",
+            "u32", "u64", "u128", "u256", "ua", "usize", "uz",
+        ];
+        let expected = vec![
+            "_ZYWX", "Z_YWX", "ZY_WX", "ZYW_X", "ZYWX", "ZYWX_", "u_zzz", "u8", "u16", "u32",
+            "u64", "u128", "u256", "ua", "usize", "uz", "v000", "v00", "v0", "v0s", "v00t", "v0u",
+            "v001", "v01", "v1", "v009", "v09", "v9", "v010", "v10", "w005s09t", "w5s009t", "x64",
+            "x86", "x86_32", "x86_64", "x86_128", "x87",
+        ];
```
I'd like to add a test cases that covers the full set of items from the example in the Style Guide, just for thoroughness:

```

    _ZYXW
    _abcd
    A2
    ABCD
    Z_YXW
    ZY_XW
    ZY_XW
    ZYXW
    ZYXW_
    a1
    abcd
    u_zzz
    u8
    u16
    u32
    u64
    u128
    u256
    ua
    usize
    uz
    v000
    v00
    v0
    v0s
    v00t
    v0u
    v001
    v01
    v1
    v009
    v09
    v9
    v010
    v10
    w005s09t
    w5s009t
    x64
    x86
    x86_32
    x86_64
    x86_128
    x87
    zyxw
```

ytmimi replied:
Looks like the example in the docs changed since I added the test. I've updated it to use the complete example from the guide.

calebcartwright commented on src/sort.rs:
```diff
@@ -0,0 +1,366 @@
+/// Represents a chunk in the version-sort algorithm
+#[derive(Debug, PartialEq, Eq)]
+enum VersionChunk<'a> {
+    /// A singel `_` in an identifier. Underscores are sorted before all other characters.
```
nit but typo

```suggestion
    /// A single `_` in an identifier. Underscores are sorted before all other characters.
```

calebcartwright commented:
This looks good to me, and I'd be good merging with the extra test case I mentioned inline (it's probably overkill but seems like a good idea to me).

Two other things I'll note while I'm thinking about them:

* the algorithm does mention spaces relative to underscores (noted below), but i presume that's the type of thing you were alluding to a follow up PR in the description?
* now that style_edition is in place, we _could_ add a standard system/idempotence test, and perhaps that would be a good way of adding a test that covers the full example mentioned in the style guide (as opposed to another unit test as i suggested inline)? 

> _ (underscore) sorts immediately after (space) but before any other character. (This ...

calebcartwright commented:
oh and I was also wondering if it would be worthwhile to try running this against some codebase that would have idents that would more thoroughly exercise the algorithm (e.g. we don't have the types of numerical, leading zeroes, etc. use cases the algorithm prescribes within the rustfmt source)

ytmimi commented:
> the algorithm does mention spaces relative to underscores (noted below), but i presume that's the type of thing you were alluding to a follow up PR in the description?

As the algorithm is implemented in this PR it only really applies to identifiers, which shouldn't be allowed to contain spaces. For that reason I didn't worry too much about space handling, and I don't think there would be scenarios where we'd need to worry about sorting things with spaces, but correct me if I'm wrong.

When it comes to follow up PRs I was mostly talking about applying the version sort to other AST nodes. I know #3764 also applied sorting to associated items, but I wasn't sure if there were others that we needed to consider as well.

---

> now that style_edition is in place, we could add a standard system/idempotence test, and perhaps that would be a good way of adding a test that covers the full example mentioned in the style guide (as opposed to another unit test as i suggested inline)?

For now I've updated the unit test with the example from the style guide, and I've also added an idempotence test in [a0f80f3](https://github.com/rust-lang/rustfmt/pull/6284/commits/a0f80f35b6d58d3e53f21556e99b4bdeeffaed67).

---

> oh and I was also wondering if it would be worthwhile to try running this against some codebase that would have idents that would more thoroughly exercise the algorithm (e.g. we don't have the types of numerical, leading zeroes, etc. use cases the algorithm prescribes within the rustfmt source)

I think this could be worthwhile. I Tried running a [Diff-Check when passing `style_edition=2024`](https://github.com/rust-lang/rustfmt/actions/runs/10675456537/job/29587463545) to see if any of the numerical ident sorting would show up, but the logs are pretty noisy, so I'm not 100% sure if that was the best idea. If you know of any crates that would be good to test against let me know, and I'd be happy to test these changes on those projects.

calebcartwright commented:
> I think this could be worthwhile. I Tried running a [Diff-Check when passing style_edition=2024](https://github.com/rust-lang/rustfmt/actions/runs/10675456537/job/29587463545) to see if any of the numerical ident sorting would show up, but the logs are pretty noisy, so I'm not 100% sure if that was the best idea. If you know of any crates that would be good to test against let me know and I'd be happy to test these changes on those projects.

Yeah that was the concern we had in our style meeting as well, that blasting it against a large, general corpus of code would be highly noisy. No need to block on this, but maybe we can get a call to action out there in some way, as it would be nice to find a good codebase that would have minimal noise
